packages/beacon-node/src/chain/blocks/verifyBlocksSignatures.ts

```diff
@@ -0,0 +1,80 @@
+import {CachedBeaconStateAllForks, getBlockSignatureSets} from "@lodestar/state-transition";
+import {allForks} from "@lodestar/types";
+import {sleep} from "@lodestar/utils";
+import {IBlsVerifier} from "../bls/index.js";
+import {BlockError, BlockErrorCode} from "../errors/blockError.js";
+import {ImportBlockOpts} from "./types.js";
+
+/**
+ * Verifies 1 or more block's signatures from a group of blocks in the same epoch.
+ * getBlockSignatureSets() guarantees to return the correct signingRoots as long as all blocks belong in the same
+ * epoch as `preState0`. Otherwise the shufflings won't be correct.
+ *
+ * Since all data is known in advance all signatures are verified at once in parallel.
+ */
+export async function verifyBlocksSignatures(
+  chain: {bls: IBlsVerifier},
+  preState0: CachedBeaconStateAllForks,
+  blocks: allForks.SignedBeaconBlock[],
+  opts: ImportBlockOpts
+): Promise<void> {
+  const isValidPromises: Promise<boolean>[] = [];
+
+  // Verifies signatures after running state transition, so all SyncCommittee signed roots are known at this point.
+  // We must ensure block.slot <= state.slot before running getAllBlockSignatureSets().
+  // NOTE: If in the future multiple blocks signatures are verified at once, all blocks must be in the same epoch
+  // so the attester and proposer shufflings are correct.
+  for (const [i, block] of blocks.entries()) {
+    // Use [i] to make clear that the index has to be correct to blame the right block below on BlockError()
+    isValidPromises[i] = opts.validSignatures
+      ? // Skip all signature verification
+        Promise.resolve(true)
+      : //
+        // Verify signatures per block to track which block is invalid
+        chain.bls.verifySignatureSets(
+          getBlockSignatureSets(preState0, block, {skipProposerSignature: opts.validProposerSignature})
+        );
+
+    // getBlockSignatureSets() takes 45ms in benchmarks for 2022Q2 mainnet blocks (100 sigs). When syncing a 32 blocks
+    // segments it will block the event loop for 1400 ms, which is too much. This sleep will allow the event loop to
+    // yield, which will cause one block's state transition to run. However, the tradeoff is okay and doesn't slow sync
+    if ((i + 1) % 8 === 0) {
+      await sleep(0);
+    }
+  }
+
+  // `rejectFirstInvalidResolveAllValid()` returns on isValid result with its index
+  const res = await rejectFirstInvalidResolveAllValid(isValidPromises);
+  if (!res.allValid) {
+    throw new BlockError(blocks[res.index], {code: BlockErrorCode.INVALID_SIGNATURE, state: preState0});
+  }
+}
+
+type AllValidRes = {allValid: true} | {allValid: false; index: number};
+
+/**
+ * From an array of promises that resolve a boolean isValid
+ * - if all valid, await all and return
+ * - if one invalid, abort immediately and return index of invalid
+ */
+export function rejectFirstInvalidResolveAllValid(isValidPromises: Promise<boolean>[]): Promise<AllValidRes> {
```

Comment:
could you move this to utils module or any utils file? this could be reused

Reply:
The counter argument is that if we stop using this function it will stay in utils forever un-used. It's very specific to this code, if it gets re-used in the future then we can move to utils